Changes to prebuild and postbuild events to download GE portable and to test/debug plugins

mast-eu opened the pull request:
Resolves #16 

## Proposed changes
* Change directory used for plugin testing and debugging to `\UserPlugins\$(ProjectName)`, because that's also the default installation dir used by the PluginManager.
* Change how the correct GE portable download is identified. The old implementation relied on `$Asset.content_type`, but in my tests I found that it doesn't work with all versions of GE, because sometimes the portable version has `$Asset.content_type=application/zip` and sometimes `$Asset.content_type=application/x-zip-compressed`.
* Some refactoring and added comments.

mast-eu commented:
Just noticed AppVeyor reports build success although the package assembly actually failed.

maraf commented:
@mast-eu Just two questions on MSBuild properties.

What is the idea behind renaming? The original approach used underscore prefix for internal properties, others were for package authors to override.

What do you mean by "shared" in names?

mast-eu commented:
> What is the idea behind renaming?

It took me time to understand some of the properties and while working on my changes I kept mixing them up. E.g. `$(GitExtensionsDownloadPath)` is a relative path and `$(_GitExtensionsDownloadPath)` is the absolute path to the same dir.
For this reason I renamed them to `$(GitExtensionsSharedRelativePath)` and `$(GitExtensionsSharedAbsolutePath)`.

> The original approach used underscore prefix for internal properties, others were for package authors to override.

I see your point now, but than it hasn't been applied consistently.
https://github.com/gitextensions/gitextensions.extensibility/blob/225bcd70a89033fd279b69a3b04177e08a31a601/src/GitExtensions.Extensibility/build/net461/GitExtensions.Extensibility.targets#L7-L9
These properties are all derived from other properties and as such should not be overridden by the plugin author.

> What do you mean by "shared" in names?

I changed the default dirname from `references` to `gitextensions.shared`, because "references" sounds very generic. On my PC, I have all repos in a common dir. It's looks something like:
```
/gitextensions
/gitextensions.*      // all the GE plugin-related repos
/GitExtensionsDoc
/SomeOtherOssProject  // not related to GE
/SomePrivateRepo      // definitely not related to GE
/references           // 🤔 , Where did this come from??
```
I think `gitextensions.shared` makes it much clearer that this dir is _related to GE_ and contains something _shared_ across multiple GE-related repos.

For the same reason I renamed all properties to `GitExtensionsShared*`. To highlight that they are _somehow_ connected to the dir `gitextensions.shared`.  Otherwise the connection is hard to grasp for plugin authors: you have to set some properties in your `.csproj` without seeing them used anywhere, because that's hidden inside a `.targets` file, which is contained in a NuGet package that you define as a dependency.

maraf commented:
Thanks for the explanation. I think that the "shared" concept is only one point of view. Plugin author can set it completely different, so it might not be shared anymore.

About property discoverability, I think they always be kind of hidden, as all MSBuild properties (it's all about that VS has intellisense for some most commonly used). I think we should rather list them in docs.

mast-eu commented:
> I think that the "shared" concept is only one point of view. Plugin author can set it completely different, so it might not be shared anymore.

I have no particular strong feelings about the property names. I just proposed names that for me seem easier to understand and remember. But if you have strong feelings about them, I'll revert to the original names without problems.

> I think we should rather list them in docs.

For sure we should document them, regardless their exact names.

maraf commented:
I think we should decide which properties are useful for override and which are not. 
I like the underscore convention. I'm not an MSBuild guru, but I have found that MS is using this convention.
Lastly, I don't see the value in adding "shared" in names. Naming is always hard and these are just opinions.

maraf commented:
`GitExtensionsDownloadPath` - public, might be relative.
`_GitExtensionsDownloadPath` - internal, always absolute.
`_GitExtensionsDownloadScriptPath` - internal, always absolute.
`GitExtensionsPath` - I thought it could be public (as read-only), as it might be usable in references (`<Reference Include="$(GitExtensionsPath)\GitUI.dll"`).
`GitExtensionsPluginsPath` - could be public (as read-only), as plugin author might use it to copy additional artifacts there.
`GitExtensionsExecutablePath` - public (as read-only), used in `launchSettings.json`.
`GitExtensionsReferenceSource` - public.
`GitExtensionsReferenceVersion` - public.

mast-eu commented:
Reading through https://github.com/gitextensions/gitextensions.extensibility/pull/11 again I understand that `GitExtensionsPath` was supposed to be a "input property" too, am I correct?

> There is also another use-case supported. In case when the user don't want to download portable GE, but he wants to use his own local build (or whatever), he can directly provide property GitExtensionsPath and the download logic will be completely skipped.

maraf commented:
> Reading through #11 again I understand that `GitExtensionsPath` was supposed to be a "input property" too, am I correct?

Oh, you are right! I completely forgot that use-case. Good catch.

RussKie commented:
It would be good to document all of these, either in the file or in a wiki

mast-eu commented:
Yes. I want to document them in the readme, or in the wiki in case it's gets too long.

maraf commented on src/GitExtensions.Extensibility/build/net461/GitExtensions.Extensibility.targets:
```diff
@@ -1,24 +1,37 @@
+        <GitExtensionsDownloadPath Condition="$(GitExtensionsDownloadPath) == ''">..\..\gitextensions.shared</GitExtensionsDownloadPath> <!-- path is relative to $(ProjectDir) -->
+        <GitExtensionsReferenceVersion Condition="$(GitExtensionsReferenceVersion) == ''">latest</GitExtensionsReferenceVersion> <!-- 'latest' or 'v3.1' (= tag from GitHub releases) or 'v3.1.0.5877' (= build number from AppVeyor)-->
+        <GitExtensionsReferenceSource Condition="$(GitExtensionsReferenceSource) == ''">GitHub</GitExtensionsReferenceSource> <!-- 'GitHub' or 'AppYevor' -->
+        <GitExtensionsPath Condition="$(GitExtensionsPath) == ''">$([System.IO.Path]::Combine('$(_GitExtensionsDownloadPath)', 'GitExtensions'))</GitExtensionsPath> <!-- for local builds (no download) -->
```
These some story about `.props` and `.targets`, if I remember it `.props` are included at the top of the project and `.targets` at the bottom. It also plays a role when user tries to use them inside the `.csproj`.

maraf replied:
I spend quite a long time with it when I was making the package to work, also with testing various setups.

mast-eu replied:
I think in my case the problem was the order of properties. In a previous version I wrote (and tested):
```
$_GitExtensionsDownloadPath = $ProjectDir + $GitExtensionsDownloadPath
$GitExtensionsPath = $_GitExtensionsDownloadPath + 'GitExtensions'
```

Then I decided to move `$GitExtensionsPath` within the same file upwards, so I got:
```
$GitExtensionsPath = $_GitExtensionsDownloadPath + 'GitExtensions'
$_GitExtensionsDownloadPath = $ProjectDir + $GitExtensionsDownloadPath
```
And this failed because  `$GitExtensionsPath` depends on `$_GitExtensionsDownloadPath`, which is only defined later.

I now changed it to:
```
$GitExtensionsPath = $ProjectDir + $GitExtensionsDownloadPath + 'GitExtensions'
$_GitExtensionsDownloadPath = $ProjectDir + $GitExtensionsDownloadPath
```
To make `$GitExtensionsPath` from `$_GitExtensionsDownloadPath` and tested it again.

maraf replied:
Silly question, why did you moved it?

mast-eu replied:
> Silly question, why did you moved it?

To have all properties grouped together rather than distributed in two different files.
> Is `_GitExtensionsDownloadPath ` used anymore?

This PR doesn't change anything in this regard. It is still the absolute path to the download dir and used in the prebuild event and passed as parameter to the PS script.

maraf replied:
Ok
